src/index.cpp

```diff
@@ -728,108 +729,88 @@ namespace diskann {
       }
       nn = Neighbor(id,
                     _distance->compare(_data + _aligned_dim * (size_t) id,
-                                       aligned_query, (unsigned) _aligned_dim),
-                    true);
+                                       aligned_query, (unsigned) _aligned_dim));
       if (fast_iterate) {
         if (inserted_into_pool_bs[id] == 0) {
           inserted_into_pool_bs[id] = 1;
-          best_L_nodes[l++] = nn;
+          best_L_nodes.insert(nn);
         }
       } else {
         if (inserted_into_pool_rs.find(id) == inserted_into_pool_rs.end()) {
           inserted_into_pool_rs.insert(id);
-          best_L_nodes[l++] = nn;
+          best_L_nodes.insert(nn);
         }
       }
       if (l == Lsize)
         break;
     }
 
     // sort best_L_nodes based on distance of each point to node_coords
-    std::sort(best_L_nodes.begin(), best_L_nodes.begin() + l);
-    unsigned best_unchanged = 0;
     uint32_t hops = 0;
     uint32_t cmps = 0;
 
-    while (best_unchanged < l) {
-      unsigned best_inserted_position = l;
```

Comment:
Before the first call to has_next() I think the NeighborSet object needs to be set to the first unchecked index. something like:
    void reset_cursor() {
      cur_ = 0;
      while (cur_ < size_ && data_[cur_].checked) {
        cur_++;
      }
    }
This would make it consistent with the way the calling code currently works. Note that calls to NeighborSet::insert() can change cur_  to a position after a checked item and has_next() could miss this item. This applies to a couple of other places in the code that walks though the NeighborSet items.

Reply:
On second look I don't think my previous comment about adding a reset_cursor() function is not needed. Your insert() function will always insert a new Neighbor object with the field checked == false and then set the 'cur_' index value to the inserted index if it's less than its current 'cur_' value. This looks OK to me.